File: heap/28_seat_reservation_manager.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Problem Description:
Manage seat reservations from 1 to n.
reserve() returns smallest available seat.
unreserve(seat) puts seat back.

Sample Input/Output:
1) reserve() -> 1
2) reserve() -> 2
3) unreserve(1), reserve() -> 1

Code Explanation:
Use min-heap of available seat numbers.
reserve pops smallest, unreserve pushes seat back.
*/

class SeatManager {
private:
    priority_queue<int, vector<int>, greater<int>> minHeap;

public:
    SeatManager(int n) {
        for (int i = 1; i <= n; i++) minHeap.push(i);
    }

    int reserve() {
        int seat = minHeap.top();
        minHeap.pop();
        return seat;
    }

    void unreserve(int seatNumber) {
        minHeap.push(seatNumber);
    }
};

int main() {
    SeatManager manager(5);
    cout << "Output 1: " << manager.reserve() << "\n";
    cout << "Output 2: " << manager.reserve() << "\n";
    manager.unreserve(1);
    cout << "Output 3: " << manager.reserve() << "\n";
    return 0;
}

File: heap/28_seat_reservation_manager.py
"""
PROBLEM: Seat Reservation Manager (LeetCode 1845)
DIFFICULTY: Medium
PATTERN: Heap - Min-Heap for available seats

DESCRIPTION:
Manage seat reservations 1 to n.
reserve() returns smallest available seat number.
unreserve(seat) marks seat as available again.

EXAMPLE:
SeatManager(5), reserve()->1, reserve()->2, unreserve(1), reserve()->1

TIME COMPLEXITY: O(log n) per operation
SPACE COMPLEXITY: O(n)
"""

import heapq


class SeatManager:
    def __init__(self, n: int) -> None:
        self.available = list(range(1, n + 1))
        heapq.heapify(self.available)

    def reserve(self) -> int:
        return heapq.heappop(self.available)

    def unreserve(self, seat_number: int) -> None:
        heapq.heappush(self.available, seat_number)


if __name__ == "__main__":
    sm = SeatManager(5)
    print(f"reserve() -> {sm.reserve()}")
    print(f"reserve() -> {sm.reserve()}")
    sm.unreserve(1)
    print(f"After unreserve(1), reserve() -> {sm.reserve()}")
    print(f"reserve() -> {sm.reserve()}")
    print(f"reserve() -> {sm.reserve()}")
